mgr/volumes: Use snapshot root directory attrs when creating clone root

ShyamsundarR opened the pull request:
If a subvolumes mode or uid/gid values are changed post a snapshot,
and a clone of a snapshot prior to the change is initiated, the clone
inherits the current source subvolumes attributes, rather than the
snapshots attributes.

Fixing this by using the snapshots subvolume root attributes to create
the clone subvolumes root.

Fixes: https://tracker.ceph.com/issues/46163

Signed-off-by: Shyamsundar Ranganathan <srangana@redhat.com>

vshankar commented:
Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77

However, ownership for subvolume root is not synchronized. It's ok to set  attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1817,8 +1837,7 @@ def test_subvolume_snapshot_clone_and_reclone(self):
 
-        # verify clone
-        self._verify_clone(subvolume, clone1)
+        # NOTE: cannot verify clone as source snapshot is removed
```
nit: should we move verify_clone before removing snapshot in all the places where possible and remove this note as it's obvious from now?

ShyamsundarR replied:
Fair enough, will do the same (add `verify_clone` and remove the snapshot post the same).

kotreshhr commented:
> Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77
> 
> However, ownership for subvolume root is not synchronized. It's ok to set attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?

We could do that. But mode is synced during creation and sync_attrs sync other attributes (uid,gid..). If we move attr sync code for root to sync_attrs, we need to make sure syncing mode is not repetitive.

vshankar commented:
> > Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77
> > However, ownership for subvolume root is not synchronized. It's ok to set attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?
> 
> We could do that. But mode is synced during creation and sync_attrs sync other attributes (uid,gid..). If we move attr sync code for root to sync_attrs, we need to make sure syncing mode is not repetitive.

You mean applying uid/gid again on root?

(BTW, IIRC, there was some request (from CSI folks) to set a custom uid/gid on a clone)

ShyamsundarR commented:
> > > Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77
> > > However, ownership for subvolume root is not synchronized. It's ok to set attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?
> > 
> > 
> > We could do that. But mode is synced during creation and sync_attrs sync other attributes (uid,gid..). If we move attr sync code for root to sync_attrs, we need to make sure syncing mode is not repetitive.
> 
> You mean applying uid/gid again on root?
> 
> (BTW, IIRC, there was some request (from CSI folks) to set a custom uid/gid on a clone)

There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
- Retaining `ceph.quota.max_bytes`
- Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
Am I missing finding these in code, or are these not set?

Further, if we extend the clone operation to specify its own custom attrs (uid/gid), it would be better coupled to set those during `clone_create`, rather than carrying them over into `async_cloner:Cloner`. (also would apply to `namespace_isolated` option if and when supported)

We should use a common method to set the attrs though, helps considering all requirements in one place (for example I did not sync the [a/m]time information). I would say `SubvolumeBase:set_attrs` is better suited to set these for root of the subvolume, thoughts?

vshankar commented:
> > > > Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77
> > > > However, ownership for subvolume root is not synchronized. It's ok to set attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?
> > > 
> > > 
> > > We could do that. But mode is synced during creation and sync_attrs sync other attributes (uid,gid..). If we move attr sync code for root to sync_attrs, we need to make sure syncing mode is not repetitive.
> > 
> > 
> > You mean applying uid/gid again on root?
> > (BTW, IIRC, there was some request (from CSI folks) to set a custom uid/gid on a clone)
> 
> There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> 
> * Retaining `ceph.quota.max_bytes`

Not retained.

> * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`

in subvolum_base::set_attrs()

>   Am I missing finding these in code, or are these not set?
> 
> Further, if we extend the clone operation to specify its own custom attrs (uid/gid), it would be better coupled to set those during `clone_create`, rather than carrying them over into `async_cloner:Cloner`. (also would apply to `namespace_isolated` option if and when supported)
> 
> We should use a common method to set the attrs though, helps considering all requirements in one place (for example I did not sync the [a/m]time information). I would say `SubvolumeBase:set_attrs` is better suited to set these for root of the subvolume, thoughts?

Fair enough. Just needs a comment that attrs for everything else is synced on clone.

kotreshhr commented:
> > > Attr fixup is done during cloning: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/async_cloner.py#L77
> > > However, ownership for subvolume root is not synchronized. It's ok to set attrs during subvolume creation -- do you think we should stick to one place for attr synchronization?
> > 
> > 
> > We could do that. But mode is synced during creation and sync_attrs sync other attributes (uid,gid..). If we move attr sync code for root to sync_attrs, we need to make sure syncing mode is not repetitive.
> 
> You mean applying uid/gid again on root?
Nope, I meant sync_attrs currently sets uid, gid, atime and mtime but 'mode' is set during file/dir creation directly. Just pointed that if we set mode in sync_attrs, it would be redundant.
> 
> (BTW, IIRC, there was some request (from CSI folks) to set a custom uid/gid on a clone)
Yes, but not sure about specific requirements. Need to check with them.

ajarr commented:
> There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> 
> * Retaining `ceph.quota.max_bytes`
> * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
>   Am I missing finding these in code, or are these not set?
> 

@ShyamsundarR do you plan to preserve the xattrs of the subvolume snapshot root as part of this PR?

ajarr commented:
> > 
> > There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> > 
> > * Retaining `ceph.quota.max_bytes`
> 
> Not retained.
> 
> > * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
> 
> in subvolum_base::set_attrs()
> 

@vshankar  I see ceph.dir.layout.pool_namespace being set on the clone root only when `--pool_layout` is passed in the `fs subvolume snapshot clone` command.

ShyamsundarR commented:
> > There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> > 
> > * Retaining `ceph.quota.max_bytes`
> > * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
> >   Am I missing finding these in code, or are these not set?
> 
> @ShyamsundarR do you plan to preserve the xattrs of the subvolume snapshot root as part of this PR?

These are the following attrs I intend to preserve/synv to the target clone subvolume as part of this PR:
- uid, gid, mode of the source subvolume snapshot
- quota xattr (`ceph.quota.max_bytes`), if any, applied from source subvolume (**NOT** snapshot)
    - As snapshot does not preserve the quota xattr, and rightly, consumes the subvolume quota
- Source subvolume snapshots pool xattr (`ceph.dir.layout.pool`), if there is no specific pool defined via args for the clone target
- If setting the pool as above, then setting the pool_namespace xattr (`ceph.dir.layout.pool_namespace`) as well on the target, if it exists, from the source subvolume snapshot

Times information (c/m)time needs to be reset post directory contents of the target are updated, as such an update would change its time information. This is already being done in the async clone [code](https://github.com/ceph/ceph/blob/749da74e2d9af075c2f365516cf9ea4370b58ec0/src/pybind/mgr/volumes/fs/async_cloner.py#L136-L140), and hence not part of the attributes being synced during the create phase.

ShyamsundarR commented:
> > > There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> > > 
> > > * Retaining `ceph.quota.max_bytes`
> > 
> > 
> > Not retained.
> > > * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
> > 
> > 
> > in subvolum_base::set_attrs()
> 
> @vshankar I see ceph.dir.layout.pool_namespace being set on the clone root only when `--pool_layout` is passed in the `fs subvolume snapshot clone` command.

We are currently passing in `isolate_namespace` as `None` when calling `set_attrs` from the create clone [method](https://github.com/ceph/ceph/blob/749da74e2d9af075c2f365516cf9ea4370b58ec0/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py#L93).

IOW, we are not setting pool namespace for a clone, at present. @ajarr I am seeing the above behavior, am I missing the something?

FWIW, a create clone currently does not take the `namespace_isolated` CLI option, so cannot create an isolated namespace when the `pool_layout` option is mentioned during clone create. @ajarr what was the driver behind the `namespace_isolated` option? We may need to add the same to the clone CLI as well based on the need for the same.

ajarr commented:
> > > > There are a couple of other items here that I see, based on these comments, that we need to handle from source root to clone root,
> > > > 
> > > > * Retaining `ceph.quota.max_bytes`
> > > 
> > > 
> > > Not retained.
> > > > * Retaining (conditionally if not being cloned to a different pool?) `ceph.dir.layout.pool_namespace`
> > > 
> > > 
> > > in subvolum_base::set_attrs()
> > 
> > 
> > @vshankar I see ceph.dir.layout.pool_namespace being set on the clone root only when `--pool_layout` is passed in the `fs subvolume snapshot clone` command.
> 

Oops! I mean to type `ceph.dir.layout.pool` in the above sentence.

> We are currently passing in `isolate_namespace` as `None` when calling `set_attrs` from the create clone [method](https://github.com/ceph/ceph/blob/749da74e2d9af075c2f365516cf9ea4370b58ec0/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py#L93).
> 
> IOW, we are not setting pool namespace for a clone, at present. @ajarr I am seeing the above behavior, am I missing the something?
> 

Nope, you aren't missing something. You're correct.

> FWIW, a create clone currently does not take the `namespace_isolated` CLI option, so cannot create an isolated namespace when the `pool_layout` option is mentioned during clone create. @ajarr what was the driver behind the `namespace_isolated` option? We may need to add the same to the clone CLI as well based on the need for the same

The idea behind that was to preserve the properties of the subvolumes created by ceph_volume_client for manila shares. The subvolumes created by ceph_volume_client by default are in a separate RADOS namespace https://github.com/ceph/ceph/blob/master/src/pybind/ceph_volume_client.py#L621
unlike those created by  `mgr volumes` interface.

Yes, we should add that `namespace_isolated` option in the clone command.

ShyamsundarR commented:
@vshankar @kotreshhr @ajarr addressed all comments, and also added additional attributes that are now synced from snapshot or subvolume root to the clone subvolume root.

I have not added the namespace option to clone in this PR, I would assume that is a separate issue/PR to address.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -88,9 +88,34 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
+            # source subvolume and snapshot attrs are used to create clone subvolume
+            # attributes of subvolume's content though, are synced during the cloning process
+            src_snappath = source_subvolume.snapshot_path(snapname)
+            st = source_subvolume.fs.stat(src_snappath)
```
nit: use statx

ShyamsundarR replied:
Addressed.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -88,9 +88,34 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
 
         subvol_path = os.path.join(self.base_path, str(uuid.uuid4()).encode('utf-8'))
         try:
+            # source subvolume and snapshot attrs are used to create clone subvolume
```
I'm thinking if it would help to move the bulk of the attribute applying stuff to subvolume base or fs_util source. The helper can accept attributes to apply so as to be reused by cloner.

ShyamsundarR replied:
Addressed.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -100,7 +100,7 @@ def load_config(self):
             self.metadata_mgr = MetadataManager(self.fs, self.config_path, 0o640)
 
-    def set_attrs(self, path, size, isolate_namespace, pool, uid, gid):
+    def set_attrs(self, path, size, isolate_namespace, pool, uid, gid, namespace=None):
```
Is there a need to pass in the additional parameter? Can't that be inferred by passing a non-None namespace?

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -100,7 +100,7 @@ def load_config(self):
             self.metadata_mgr = MetadataManager(self.fs, self.config_path, 0o640)
 
-    def set_attrs(self, path, size, isolate_namespace, pool, uid, gid):
+    def set_attrs(self, path, size, isolate_namespace, pool, uid, gid, namespace=None):
```
Is there a need to pass in an additional parameter? Can't that be inferred by passing a non-None namespace?

ShyamsundarR replied:
Meaning, we use `isolate_namespace` itself to pass the value or generate one instead? Or, we generate one as needed before the call to `set_attrs` and then pass it along in `namespace` and remove `isolate_namespace`?

vshankar replied:
just pass namespace -- if it's invalid (None), it need not be set.

ShyamsundarR replied:
Addressed.

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -88,9 +88,34 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
+
+            # apply source subvolume size restrictions to the clone
+            try:
+                size = int(source_subvolume.fs.getxattr(source_subvolume.path, 'ceph.quota.max_bytes').decode('utf-8'))
```
If source subvolume size restrictions are applied to the clone, won't the clone fail in following scenario.

1. Create subvolume of size 1GB (assume 1GB is written)
2. Take a subvolume snapshot
3. Delete 500 MB and resize the subvolume to 500 MB
4. Clone the subvolume snapshot

ajarr replied:
Good point. I tried this out. As expected the clone status of the original snapshot remains in progress, and I see the following traceback,
```
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/fs_util.py", line 117, in copy_file
    written += fs.write(dst_fd, data[written:], -1)
  File "cephfs.pyx", line 1463, in cephfs.LibCephFS.write
cephfs.Error: error in write: Disk quota exceeded [Errno 122]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_job.py", line 44, in run
    self.async_job.execute_job(vol_job[0], vol_job[1], should_cancel=lambda: thread_id.should_cancel())
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 309, in execute_job
    clone(self.vc, volname, job[0].decode('utf-8'), job[1].decode('utf-8'), self.state_table, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 222, in clone
    start_clone_sm(volume_client, volname, index, groupname, subvolname, state_table, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 202, in start_clone_sm
    (next_state, finished) = handler(volume_client, volname, index, groupname, subvolname, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 159, in handle_clone_in_progress
    do_clone(volume_client, volname, groupname, subvolname, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 155, in do_clone
    bulk_copy(fs_handle, src_path, dst_path, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 144, in bulk_copy
    cptree(source_path, dst_path)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 129, in cptree
    copy_file(fs_handle, d_full_src, d_full_dst, mo, cancel_check=should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/fs_util.py", line 120, in copy_file
    raise VolumeException(-e.args[0], e.args[1])
TypeError: bad operand type for unary -: 'str'
```

ajarr replied:
I expected the snapshot to preserve the xattr `ceph.quota.max_bytes` like it does for `ceph.dir.layout`, but it doesn't it. Not sure why.

vshankar replied:
> Good point. I tried this out. As expected the `fs clone status` of the original snapshot's clone remains in progress, and I see the following traceback,

Clone status should be "failed" -- guess the async cloner hit an unhandled exception here.

So, the quota xattr is not preserved in the snapshot -- that seems like a bug.

vshankar replied:
The xattr should be synced from the snapshot anyway...

ShyamsundarR replied:
The quota xattr applies to the subvolume and the size occupied by its snapshots as well, I would assume.

IOW, creating content -> taking a snapshot -> deleting content, would allow a user to exceed the quota restriction on the subvolume, and store content in the snapshots instead.

Considering this, I think the manner of solving this would be to store the quota at the time of snapshot in the metadata file, like we intend to do for snapshot df content.

ShyamsundarR replied:
> The quota xattr applies to the subvolume and the size occupied by its snapshots as well, I would assume.
> 
> IOW, creating content -> taking a snapshot -> deleting content, would allow a user to exceed the quota restriction on the subvolume, and store content in the snapshots instead.

Tested the above, and found out that snapshot COW blocks are not accounted for in the directory quota. Other results present in https://tracker.ceph.com/issues/46278 

> 
> Considering this, I think the manner of solving this would be to store the quota at the time of snapshot in the metadata file, like we intend to do for snapshot df content.

With the observations in the tracker, this can be solved fixing the xattr being available in the snapshots, or for subvolumes storing this in the meta as mentioned above.

ajarr replied:
> > Good point. I tried this out. As expected the `fs clone status` of the original snapshot's clone remains in progress, and I see the following traceback,
> 
> Clone status should be "failed" -- guess the async cloner hit an unhandled exception here.
> 
> So, the quota xattr is not preserved in the snapshot -- that seems like a bug.

https://tracker.ceph.com/issues/46360
Also posted fix

ShyamsundarR replied:
Addressed picking up quota from the snapshot as related tracker https://tracker.ceph.com/issues/46278 is now fixed.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -100,33 +101,62 @@ def load_config(self):
+        try:
+            attrs["data_pool"] = self.fs.getxattr(pathname, 'ceph.dir.layout.pool').decode('utf-8')
+        except cephfs.NoData:
+            attrs["data_pool"] = None
```
I am not sure how to tackle the mypy errors in these lines, dropping it here if others have any thoughts. Causes failure in the Jenkins make check job.

```
volumes/fs/operations/versions/subvolume_base.py:118: error: Incompatible types in assignment (expression has type "None", target has type "int")
volumes/fs/operations/versions/subvolume_base.py:123: error: Incompatible types in assignment (expression has type "None", target has type "int")
volumes/fs/operations/versions/subvolume_base.py:128: error: Incompatible types in assignment (expression has type "None", target has type "int")
```

kotreshhr replied:
Initialize attrs["data_pool"] = None before usage. This way the type is set to "Any" and mypy wouldn't complain.

Reference:
1. [mypy-common-issues](https://mypy.readthedocs.io/en/stable/common_issues.html#redefinitions-with-incompatible-types)
2. [Any-type](https://docs.python.org/3/library/typing.html)

ajarr replied:
Maybe go with option 2 in https://stackoverflow.com/questions/43910979/mypy-error-incompatible-types-in-assignment
So something like,
```
attrs = {}  # type: Dict[str, Union[int, str, None]]
```

ShyamsundarR replied:
Thanks @kotreshhr @ajarr 

The mypy error is assuming the entire dict stores [str, int], as the first element that I added was an int. Hence the suggestion to initialize it to a None type did not work.

The solution to add a type comment instead worked fine and learnt from @ajarr that this was legal from a [PEP](https://www.python.org/dev/peps/pep-0484/#type-comments) standpoint as well.

kotreshhr replied:
:+1:

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -45,7 +45,13 @@ def create(self, size, isolate_nspace, pool, mode, uid, gid):
             # create directory and set attributes
             self.fs.mkdirs(subvol_path, mode)
-            self.set_attrs(subvol_path, size, isolate_nspace, pool, uid, gid)
+            attrs = {}
```
nit: this could just be:

```
attrs = {
'uid' : ...
...
...
}
```

ShyamsundarR replied:
Addressed

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -100,33 +101,62 @@ def load_config(self):
+
+        return attrs
+
+    def set_attrs(self, path, attrs):
```
> Do you plan to call this in cloner?

Not in this PR or as yet. Just creating a cleaner interface as you suggested, for now.

vshankar replied:
+1 if it can be included in this PR.

ShyamsundarR replied:
> general comment: pythonic way of writing this would be:
> 
> ```
> if attrs.get(key, None):
>    ....
> ```
> 
> this way, the routine can handle a subset of all supported attrs.

Addressed the above.

ShyamsundarR replied:
> +1 if it can be included in this PR.

@vshankar I intend to pick this up when addressing https://tracker.ceph.com/issues/46399 (preserve xattrs during snapshot clone), which is about 2 issues down the line in my queue.

As at that point in time we would add get/set xattrs as well to these functions and use the common methods in both setting up the subvolume root and any directories and files within it.

Let me know if this works for you.

vshankar replied:
ack -- works for me.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -88,9 +95,20 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
 
         subvol_path = os.path.join(self.base_path, str(uuid.uuid4()).encode('utf-8'))
         try:
+            # source subvolume and snapshot attrs are used to create clone subvolume.
```
Addressed, it was in a limbo w.r.t quota from subvolume, but that is now fixed.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -92,8 +96,28 @@ def _verify_clone_attrs(self, subvolume, clone, source_group=None, clone_group=N
+        if clone_pool:
+            # verify pool is set as per request
+            clone_path = self._get_subvolume_path(self.volname, clone)
+            ondisk_pool = self.mount_a.getfattr(clone_path, "ceph.dir.layout.pool")
```
Addressed.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -92,8 +96,28 @@ def _verify_clone_attrs(self, subvolume, clone, source_group=None, clone_group=N
 
-    def _verify_clone(self, subvolume, clone, source_group=None, clone_group=None, timo=120):
-        path1 = self._get_subvolume_path(self.volname, subvolume, group_name=source_group)
+    def _verify_clone_root(self, subvolume, snapshot, clone, source_group, clone_group, clone_pool):
```
Added a comment to explain what is verified here.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -92,8 +96,28 @@ def _verify_clone_attrs(self, subvolume, clone, source_group=None, clone_group=N
+            source_path = self._get_subvolume_snapshot_path(subvolume, snapshot, source_group)
+            self.assertEqual(clone_info["data_pool"],
+                             self.mount_a.getfattr(source_path, "ceph.dir.layout.pool"))
+            self.assertEqual(clone_info["pool_namespace"],
```
Namespace (and pool) is inherited from snapshot if pool is not set explicitly, otherwise namespace is not inherited or set for a clone, only the pool attribute is set. Hence this is not compared in the common part of the if..else structure.

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -102,17 +108,37 @@ def _verify_clone_attrs(self, subvolume, clone, source_group=None, clone_group=N
-        if snapshot and subvol_path:
-            path1 = self._get_subvolume_snapshot_path(snapshot, subvol_path)
+    def _verify_clone_root(self, source_path, clone_path, clone, clone_group, clone_pool):
+        # verifies clone root attrs that are explicitly set or inherited from the snapshot it was
```
can't you just say that only quota, pool layout, pool namespace at the clone root is checked here?

ShyamsundarR replied:
Addressed.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -120,9 +127,18 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
+            # override snapshot pool setting, if one is provided for the clone
+            if pool is not None:
+                attrs["data_pool"] = pool
+                attrs["pool_namespace"] = None
```
Yes, my understanding is that a clone will (in the future) have a `namespace_isolated` parameter to force isolation, when changing pools.

If that is not the case then we need to change it to use the same namespace as the clone source.

ajarr replied:
OK. Makes sense that we add that `namespace_isolated` parameter to `fs subvolume snapshot` clone command.

ShyamsundarR commented:
@vshankar requesting a review, thanks.

vshankar reviewed (approved):
LGTM

ajarr commented:
http://pulpito.front.sepia.ceph.com/rraja-2020-08-19_13:10:02-fs-ajarr-wip-46163-distro-basic-smithi/

$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=ajarr-wip-46163
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH  --filter "volumes" -N 5

ajarr commented:
> http://pulpito.front.sepia.ceph.com/rraja-2020-08-19_13:10:02-fs-ajarr-wip-46163-distro-basic-smithi/
> 
> $ export CEPH_QA_MAIL="[rraja@redhat.com](mailto:rraja@redhat.com)"; MACHINE_NAME=smithi; CEPH_BRANCH=ajarr-wip-46163
> $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes" -N 5


All the upgrade volume jobs failed. There are 2 failed test_volumes jobs as well. Need to dig further.

ShyamsundarR commented:
> > http://pulpito.front.sepia.ceph.com/rraja-2020-08-19_13:10:02-fs-ajarr-wip-46163-distro-basic-smithi/
> > $ export CEPH_QA_MAIL="[rraja@redhat.com](mailto:rraja@redhat.com)"; MACHINE_NAME=smithi; CEPH_BRANCH=ajarr-wip-46163
> > $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes" -N 5
> 
> All the upgrade volume jobs failed. There are 2 failed test_volumes jobs as well. Need to dig further.

Looked at the test_volumes failures and here is what I see: (more details [here](https://gist.github.com/ShyamsundarR/0a227f9d5f275482edc0f114b6930e0a))

- Failure in both test cases relate to the command from the test timing out
- On the mgr logs post recieving the command the logs indicate the following,
```
2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes INFO volumes.module] Starting _cmd_fs_subvolume_getpath(prefix:fs subvolume getpath, sub_name:subvolume_0000000000545095, target:['mon-mgr', ''], vol_name:cephfs) < ""
2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes DEBUG mgr_util] self.fs_id=46, fs_id=47
2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes WARNING mgr_util] filesystem id changed for volume 'cephfs', reconnecting...
2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes DEBUG mgr_util] self.fs_id=46, fs_id=47
2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes INFO mgr_util] aborting connection from cephfs 'cephfs'
2020-08-19T16:30:55.586+0000 7f43fcf9f700  2 client.16292 unmounting (abort)
2020-08-19T16:30:55.586+0000 7f43fcf9f700  5 client.16292 _closed_mds_session mds.0 seq 219
```

ShyamsundarR commented:
> > > http://pulpito.front.sepia.ceph.com/rraja-2020-08-19_13:10:02-fs-ajarr-wip-46163-distro-basic-smithi/
> > > $ export CEPH_QA_MAIL="[rraja@redhat.com](mailto:rraja@redhat.com)"; MACHINE_NAME=smithi; CEPH_BRANCH=ajarr-wip-46163
> > > $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes" -N 5
> > 
> > 
> > All the upgrade volume jobs failed. There are 2 failed test_volumes jobs as well. Need to dig further.
> 
> Looked at the test_volumes failures and here is what I see: (more details [here](https://gist.github.com/ShyamsundarR/0a227f9d5f275482edc0f114b6930e0a))
> 
>     * Failure in both test cases relate to the command from the test timing out
> 
>     * On the mgr logs post recieving the command the logs indicate the following,
> 
> 
> ```
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes INFO volumes.module] Starting _cmd_fs_subvolume_getpath(prefix:fs subvolume getpath, sub_name:subvolume_0000000000545095, target:['mon-mgr', ''], vol_name:cephfs) < ""
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes DEBUG mgr_util] self.fs_id=46, fs_id=47
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes WARNING mgr_util] filesystem id changed for volume 'cephfs', reconnecting...
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes DEBUG mgr_util] self.fs_id=46, fs_id=47
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes INFO mgr_util] aborting connection from cephfs 'cephfs'
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  2 client.16292 unmounting (abort)
> 2020-08-19T16:30:55.586+0000 7f43fcf9f700  5 client.16292 _closed_mds_session mds.0 seq 219
> ```

In other cases where the fs_id changes, re-connections seem to complete within a second. In the failure case it does not complete (in >120 seconds). Also, noted the following log messages during the failure which look significant,

_preceded by line number in the log_
```
720953 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes WARNING mgr_util] filesystem id changed for volume 'cephfs', reconnecting...
720954 2020-08-19T16:30:55.586+0000 7f43fcf9f700  0 [volumes DEBUG mgr_util] self.fs_id=46, fs_id=47
<snip>
721010 2020-08-19T16:30:55.590+0000 7f43fcf9f700  2 client.16292 cache still has 0+2 items, waiting (for caps to release?)
<snip>
721952 2020-08-19T16:31:00.590+0000 7f43fcf9f700  1 client.16292 dump_cache
 721953 2020-08-19T16:31:00.590+0000 7f43fcf9f700  1 client.16292 dump_inode: DISCONNECTED inode 0x10000000003 #0x10000000003 ref 1 0x10000000003.head(faked_ino=0 ref=1 ll_ref=0 cap_refs={} open={} mode=40755 size=0/0 nlink=0 btime=0.0000        00 mtime=2020-08-19T16:30:30.921707+0000 ctime=2020-08-19T16:30:30.925234+0000 caps=- has_dir_layout quota(max_bytes = 10485760 max_files = 0) 0x56243f16b080)
 721954 2020-08-19T16:31:00.590+0000 7f43fcf9f700  1 client.16292 dump_inode: DISCONNECTED inode 0x10000000003 #0x10000000003 ref 1 0x10000000003.snapdir(faked_ino=0 ref=1 ll_ref=0 cap_refs={} open={} mode=40755 size=0/0 nlink=1 btime=202        0-08-19T16:30:15.794687+0000 mtime=2020-08-19T16:30:15.794687+0000 ctime=2020-08-19T16:30:15.999257+0000 caps=- 0x562444cff700)
 721955 2020-08-19T16:31:00.590+0000 7f43fcf9f700  2 client.16292 cache still has 0+2 items, waiting (for caps to release?)
```

lxbsz commented:
> Looks like a duplicate of https://tracker.ceph.com/issues/46882 (which is in progress)

Yeah, it is. It dues to there has the Inode ref leak issue, the fixing PR is https://github.com/ceph/ceph/pull/36580.

Thanks.

ajarr commented:
> > http://pulpito.front.sepia.ceph.com/rraja-2020-08-19_13:10:02-fs-ajarr-wip-46163-distro-basic-smithi/
> > $ export CEPH_QA_MAIL="[rraja@redhat.com](mailto:rraja@redhat.com)"; MACHINE_NAME=smithi; CEPH_BRANCH=ajarr-wip-46163
> > $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter "volumes" -N 5
> 
> All the upgrade volume jobs failed. There are 2 failed test_volumes jobs as well. Need to dig further.

I've created a tracker ticket for the failure in upgrade volume jobs,
https://tracker.ceph.com/issues/47051

ajarr commented:
jenkins test api

ajarr commented:
The jenkins API test failure is https://tracker.ceph.com/issues/46848
